src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/VTableSliceNode.cs

```diff
@@ -121,9 +118,25 @@ public override IReadOnlyList<MethodDesc> Slots
         {
             get
             {
-#if DEBUG
-                _slotsCommitted = true;
-#endif
+                if (_slots == null)
+                {
+                    // Sort the lazily populated slots in metadata order.
+                    // This ensures that Foo<string> and Foo<object> will end up with the same vtable
+                    // no matter the order in which VirtualMethodUse nodes populated it.
+                    ArrayBuilder<MethodDesc> slotsBuilder = new ArrayBuilder<MethodDesc>();
+                    DefType defType = _type.GetClosestDefType();
+                    foreach (var method in defType.GetAllMethods())
+                    {
+                        if (_usedMethods.Contains(method))
+                            slotsBuilder.Add(method);
+                    }
+                    Debug.Assert(_usedMethods.Count == slotsBuilder.Count);
+                    _slots = slotsBuilder.ToArray();
+
+                    // Null out used methods so that we AV if someone tries to add now.
+                    _usedMethods = null;
```

Comment:
An assert in `AddEntry` that _usedMethods != null would make the intent clearer for future late adders of virtual methods.